### ISSUE TITLE
Allow to enforce a minimum Quarkus Core version for a given extension version

### ISSUE THREAD

**Issue by ia3andy** [quarkusio/quarkus#37169]
### Description

Currently there is a notion of the version of Quarkus core used for building an extension (and it will reflect on the tooling).

Users can still easily change the version of Quarkus or extensions in the pom.xml leading to possible issues: No class def found, Methods, ..., but also sometime it may build and work but without critical features (security, ...).

It is important to provide a proper way to enforce when releasing an extension that user apps will be built using a compatible version of Quarkus.

### Implementation ideas

- A flag to specify that the build with quarkus core version needs to be enforced as minimum.
- Or make it the default behavior and make sure we don't update Quarkus Core when it's not required.

cc @aloubyansky @gastaldi @max

**Comment by maxandersen** [comment]
I'm failing to see how this is specific to quarkus - this can happen for literally *any* jar dependency and a single extension saying "No, do NOT build with a different version" is not going to work well.

"make sure we don't update Quarkus Core version in extensions when it's not required." is exactly what we recommend
 - its how one stay compatible with things.

**Comment by maxandersen** [comment]
btw. qarkus update and quarkus info is where we have ways to help/warn the user today.

**Comment by ia3andy** [comment]
> I'm failing to see how this is specific to quarkus - this can happen for literally _any_ jar dependency and a single extension saying "No, do NOT build with a different version" is not going to work well.
> 
> "make sure we don't update Quarkus Core version in extensions when it's not required." is exactly what we recommend
> 
> * its how one stay compatible with things.

What I mean is to say "you need to upgrade the Quarkus version for it to be compatible with this extension version". This is not what you are saying I think.

I am not talking about locking to a specific version of Quarkus but having a minimum version to enforce. 

I don't see any reason why we should keep building user projects when we know it will fail and be incompatible! We have the possibility to enforce it, why not do it.

**Comment by ia3andy** [comment]
Today I released Quarkus Web Bundler 1.3.0 which is compatible with Quarkus 3.7.3, with a lower Quarkus version it will build but there might be issues really hard to catch. 

This kind of scenario is really bad for users as there is no way to warn them about this incompatibility.

**Comment by gastaldi** [comment]
How about introducing a Build Item (`MinimumQuarkusVersionBuildItem`) that extensions could produce and create a consumer in core that checks whether the minimum Quarkus version is in range and fails the build if not?

**Comment by ia3andy** [comment]
It would be nice if we could backport it on previous versions 🤣

**Comment by gastaldi** [comment]
We could have that as a separate extension and move to core later on. Then people would just need to add this extra extension to perform the check

**Comment by ia3andy** [comment]
the question is, do we have access to the Quarkus version?

**Comment by aloubyansky** [comment]
> Currently there is a notion of the version of Quarkus core used for building an extension 

I guess you meant building an application?

**Comment by dmlloyd** [comment]
> How about introducing a Build Item (`MinimumQuarkusVersionBuildItem`)

Does it have to be a build item?

**Comment by gastaldi** [comment]
> Does it have to be a build item?

Not necessarily, just wanted an API that extensions could use to validate the deployment

**Comment by ia3andy** [comment]
@aloubyansky @gastaldi this is really problematic, could we provide a solution?

**Comment by ia3andy** [comment]
(Even more with the config update that breaks it all)

**Comment by aloubyansky** [comment]
We should be able to perform this check before we build now. I'll have a look.

**Comment by ia3andy** [comment]
We should start the workgroup for this based on what we decided at the f2f. @holly-cummins do you have the pictures?

**Comment by aloubyansky** [comment]
From the last discussion we had, we could check for the lower boundary Quarkus version that is known at extension release time but we wouldn't be able to check compatibility from that point on w/o querying the registry (assuming the registry gets up-to-date info about Quarkus core compatibility).
